### PR TITLE
fix(Config): Ensure that invalid user config is handled gracefully

### PR DESCRIPTION
**Pull request by pgrzesik**
I've figured it might be a bit easier to discuss with a rough implementation at hand, as of right now I'm not sure about a few things and I'm hoping you'd be able to help out with it:

1. Looking at current convention for logging errors, it seems like any error logging for users happen if `SLS_DEBUG` is set - should all added logging + file rename also happen only if that var is set? In addition, what is the preferred convention when it comes to logging errors from this module? I've started with the same approach as we already use, but I'm not sure if it's still the preferred option. 
2. Refactoring to async - I've noticed that we're currently moving away from sync and promises in favor of `async/await` syntax - whole config module is currently using `sync` fs operations. Do you think it would make sense to start introducing `async` approach in this module as well? The only downside here is that it might trigger a "domino effect" in other projects where in order to accomodate to new syntax, we  might need to refactor all code that is using this module to also adopt `async/await` syntax. So far I've noticed that `serverless/serverless` and `serverless/components` are using it, not sure about other projects.
3. As for renaming invalid config file to `.bak` and generating fresh config. I've tested `rc` and in cases of e.g. malformed JSON, they don't throw any meaningful error other than JSON parsing error, they also don't return a reference to specific file that contained invalid syntax. Internally, `rc` checks if content starts with `{` and if that's the case, it is assumed that it's JSON, which makes it hard to actually locate the problematic file. In addition, the only config file that is automatically generated by use is the global one, I'm not sure about renaming and generating a new local config file, because it would have to be pretty much the same as a global one which makes it redundant - hence the suggestion to only rename and regenerate the global config file. What is your opinion on that?

Closes: #61

### CODE REVIEW

File: config.js [hunk 1]
```diff
@@ -83,13 +83,38 @@ function getConfig() {
...
+  // TODO: TESTS
+  try {
+    return rc(rcFileBase, null, /* Ensure to not read options from CLI */ {});
+  } catch (rcError) {
+    process.stdout.write(`${chalk.red(rcError.stack)}\n`);
+    // TODO: IMPROVE ERR MESSAGE
+    process.stdout.write(
+      `Serverless: ${chalk.red(`Unable to read config due to ${rcError.message} error`)}\n`
+    );
+    return {};
```
Comment (pgrzesik) [inline config.js:95]: Here, we could also once again try to get and return the global config via `getGlobalConfig` - it would allow us to ensure that if the global config was the problematic one, it was renamed and regenerated.

**Comment by medikoo** [comment]
@pgrzesik very good questions

> what is the preferred convention when it comes to logging errors from this module

Actually it wasn't coined so far. 

So far in a Framework logs handling is quite primitive, we have general (always on) logs, and some more verbose logging (which provide more information that's helpful for debugging purposes) which can be turned on with `SLS_DEBUG=1`.

This is scheduled to be vastly improved by addressing https://github.com/serverless/serverless/issues/1720 (which has already proposed solution) - still that should not block us here.

For context of this util. I suggest that we first address https://github.com/serverless/serverless/issues/8444, but move the logging util to `log.js` in this package (and not to `lib/utils/log.js` as proposed in an issue).

Then for logging we can use that util here, and I'd log file syntax errors unconditionally as warnings (we color them in orange as here: https://github.com/serverless/serverless/blob/e32b7714253108f9078d2218e68c5994f20cde64/lib/classes/ConfigSchemaHandler/index.js#L78-L83)

> Do you think it would make sense to start introducing async approach in this module as well? 

Totally, and it can be done in all packages which were already upgraded to drop support for Node.js v6 (and that's all of them, apart `@serverless/test` which have that coming very soon with this PR: https://github.com/serverless/test/pull/59).

Here we can use it right away -> https://github.com/serverless/utils/blob/3234bdbd6a5e5b559757fa1cec670ffe9793393d/package.json#L101

> The only downside here is that it might trigger a "domino effect" 

It shouldn't, as async/await refactor does not change anything for consumer module (apart of fact that they also should have already dropped support for Node.js v6). 

Currently we refactor, on the go, when needed (trying to refactor everything, would be to big task to jump into). Also it's great to not mix such refactors with introduction of new functionalities. If we see that we could benefit from from async/await refactor it might be nice to first propose PR just with that refactor, and then follow with the functionality PR (although in our case where we merge via rebase, it's fine to split this work with individual commits in single PR)

Do you see such need here? I think all `config.js` functions are _sync_, and I'm not sure about changing that, as I think there's a reason in it. (e.g. it's fine to schedule writes sync way, without confirming that they went successful.. and in case of reads, it should I believe be read once and then accessed from cache)

> I'm not sure about renaming and generating a new local config file, because it would have to be pretty much the same as a global one which makes it redundant - 

In that case let's just show warning, and assume no config exists (so override errorneous one without backing it up).
It's really an edge case and most likely to be caused by manual play with config content.

**Comment by pgrzesik** [comment]
Thanks @medikoo for answering all my questions. I've prepared the updated version, but I'd still would like to get your opinion on a few things, I've added comments directly on lines that I'm not sure about

File: config.js (hunk 1 above)
Comment (pgrzesik) [reply config.js:86]: As mentioned above, here we have a bit tricky situation at the moment. With the current behavior, if someone uses `getConfig` and `rc` will throw an error, we're not renaming or generating new global config file, but rather issue a warning and return an empty config. At this point we're not sure if local or global config file was problematic. What we could potentially do is after catching the error, instead of `return {}`, we could still try to return the global config, via `return getGlobalConfig()`. What that would give us is the ability to rename and regenerate global config if it was in fact the one causing troubles. What is your opinion on that ?

File: config.js [hunk 2]
```diff
@@ -83,13 +79,33 @@ function getConfig() {
     createConfig();
   }
   // then return config merged via rc module
-
-  return rc(rcFileBase, null, /* Ensure to not read options from CLI */ {});
+  try {
+    return rc(rcFileBase, null, /* Ensure to not read options from CLI */ {});
+  } catch (rcError) {
+    log(`Unable to read config due to ${rcError.message} error`, { color: 'orange' });
```
Comment (pgrzesik) [inline config.js:85]: Do you think we should also log `error.stack` in these situations? I believe we shouldn't do that for warnings, but wanted to get your opinion. (or maybe log them if `SLS_DEBUG` is set?)

File: test/config.js [hunk 3]
```diff
@@ -52,3 +54,52 @@ describe('config', () => {
...
+      (data) => (stdoutData += data),
+      () => {
+        conf = config.getGlobalConfig();
+      }
+    );
+
+    const backupConfigFilePath = `${config.CONFIG_FILE_PATH}.bak`;
+
+    expect(conf).to.not.be.empty;
+    expect(stdoutData).to.include('Unable to read global config');
```
Comment (pgrzesik) [inline test/config.js:92]: This is kinda iffy approach to testing logging, but I didn't come up with a better version that is able to mock/spy the `log` util with `sinon` - I'd love to hear your idea if you have a suggestion on how to improve it :100:

File: config.js (hunk 1 above)
Comment (medikoo) [reply config.js:86]: > What we could potentially do is after catching the error, instead of return {}, we could still try to return the global config, via return getGlobalConfig().

I think it's a very good call, and in such case we would never override faulty config without backing it up first. I'd go for it

File: config.js (hunk 2 above)
Comment (medikoo) [reply config.js:85]: I think full stacks traces should apply only with `SLS_DEBUG`

Do you think in this case full stack trace would be helpful? (From those stacks you've observed will it give an additional aid exposing them in full ?) If not, I would not consider it. If yes, we may add that if `SLS_DEBUG` is set.

Also I would reword warning slightly:

```
User Configuration error: Cannot resolve config file: ${error.message}`
```

Comment (medikoo) [reply config.js:85]: Also I wouldn't show warning or output anything if we redirect to global read, just let global read handle eventual error. That should make processing easier

File: test/config.js (hunk 3 above)
Comment (medikoo) [reply test/config.js:92]: TBH I would just test, that util doesn't crash on errorneous file. I wouldn't confirm on warning logs output (we usually don't do that, as it's usually not involved in functionality outcome, and there are no good means to test human readable messages which are subject to constant improvements)

File: config.js [hunk 4]
```diff
@@ -83,13 +79,33 @@ function getConfig() {
...
+      try {
+        // try/catch to account for very unlikely race condition where file existed
+        // during hasConfigFile check but no longer exists during rename
+        const backupServerlessrcPath = `${serverlessrcPath}.bak`;
+        fs.renameSync(serverlessrcPath, backupServerlessrcPath);
+        log(
+          `Your previous config was renamed to ${backupServerlessrcPath} for debugging. Default global config will be recreated under ${serverlessrcPath}.`,
+          { color: 'orange' }
+        );
+      } catch (renameError) {
```
Comment (medikoo) [inline config.js:105]: We may skip error argument

File: config.js (hunk 2 above)
Comment (pgrzesik) [reply config.js:85]: I've tested it out and the stack traces are not really helpful so I've decided against including them. If there's a problem with accessing the file by `rc`, the file will be skipped silently, the only error that will bubble up is the JSON parsing error, but the stack trace doesn't add any value in this case as the error message already explains what happened

Comment (medikoo) [reply config.js]: > I've tested it out and the stack traces are not really helpful so I've decided against including them.

How exactly message with a stack trace looks?

File: config.js [hunk 5]
```diff
@@ -83,13 +79,37 @@ function getConfig() {
...
   }
   // then return config merged via rc module
-
-  return rc(rcFileBase, null, /* Ensure to not read options from CLI */ {});
+  try {
+    return rc(rcFileBase, null, /* Ensure to not read options from CLI */ {});
+  } catch (rcError) {
+    log(`User Configuration error: Cannot resolve config file: ${rcError.message}`, {
+      color: 'orange',
+    });
```
Comment (medikoo) [inline config.js:87]: Let's maybe reword `error` to `warning` (as we do not fail, and logically we treat it as warning).

I remember when first we were reporting configuration validation errors as _errors_ and most of the users assumed things got broken

**Review by medikoo — CHANGES_REQUESTED** [review]
Looks great @pgrzesik ! I have just one minor suggestion towards warnings wording

File: config.js (hunk 2 above)
Comment (medikoo) [reply config.js]: @pgrzesik Ignore above message (I've read _I've tested it out and the stack traces are really helpful_)

**Review by medikoo — APPROVED** [review]
Looks great! 👍
